PG-1419 Validate key provider on creation

AndersAstrand opened the pull request:
The validation done is very rudimentary for now. The main part of this PR is some refactoring of tde_keyring.c. I'm not fully happy with where it's at, but I think I like it better than what was there at least.

~I did leave a TODO in there for any reviewers to comment on! The comment will ofc not be merged, I'll either remove the comment or change the behaviour.~

Commit:
Change some keyring function to return void

No callers cared about the return value of these functions anyway.

Commit:
Refactor some lower level functions in tde_keyring

Previously write_key_provider_info() was a bit of a "do everything"
function that had very different behavior depending on what parameters
was passed to it. This commit reworks it to a "dumb" function that just
writes the data without asking questions and have the callers take
responsibility for data validity.

This is to make it easier to validate the data in different ways
depending on the caller's needs without further complicating
write_key_provider_info().

Commit:
PG-1419 Validate key provider access

This adds some validation to make sure we can access the key provider
when it's created to make the user experience a little nicer. The actual
access validation is very rudimentary for now but can easily be
expanded.

Commit:
Remove non-test

This regression file didn't really test anything. The tests it was
supposed to do was removed here https://github.com/percona/pg_tde/commit/e270322f72f80dfd98f951f30400e270a5d41cee
So now it doesn't test anything that key_provider.sql doesn't already do
for us.

dutow commented on contrib/pg_tde/sql/vault_v2_test.sql:
```diff
@@ -31,4 +31,7 @@ SELECT pg_tde_verify_key();
 
 DROP TABLE test_enc;
 
+-- Creating provider fails if we can't connect to vault
```
connect to kmip

AndersAstrand replied:
No this is the vault test i think :)

dutow replied:
Ops, right, I somehow mixed up the two files :)

AndersAstrand commented:
I don't know why the "Test PSP with TDE" job fails. Will look into that.

dutow commented on contrib/pg_tde/src/include/catalog/tde_keyring.h:
```diff
@@ -22,24 +22,25 @@ typedef struct KeyringProviderRecord
 	off_t		offset_in_file;
 	KeyringProviderRecord provider;
-} KeyringProviderXLRecord;
+} KeyringProviderFileRecord;
```
This name is misleading, at first I thought it's related to keyring file.

AndersAstrand replied:
I agree that the name is bad. Will think about it some more.

AndersAstrand replied:
Maybe `KeyringProviderRecordInFile` ?

Regardless of naming i will move the renaming to a separate commit, because there really is no need to have it in the same commit as the other changes.

jeltz replied:
While I am not the biggest fan of it maybe we should follow the pattern of what PostgreSQL does, e.g. `xl_btree_newroot` and `xl_smgr_create` which would make this struct called `xl_pgtde_writekeyprovider` or `xl_pg_tde_write_key_provider`.

But I think really this should be a part of a separate pr where we pick consistent names for all xlog records.

AndersAstrand replied:
Are you also suggesting that we shouldn't use this struct for other things that are not WAL related? I used it because it was a convenient way to pass around the data with it's location in file and such attached, but maybe I should just ditch that and use separate variables for it?

jeltz replied:
Hmm, good point. No idea then really. Do as you feel good.

dutow replied:
`KeyringProviderRecordInFile` sounds better.

dutow commented on contrib/pg_tde/src/keyring/keyring_file.c:
```diff
@@ -143,3 +145,19 @@ set_key_by_name(GenericKeyring *keyring, KeyInfo *key)
 }
+
+static void
+validate(GenericKeyring *keyring)
```
I still think the file keyring should do more validation/restrictions, that will be a different pr?

AndersAstrand replied:
If you're talking about restricting where we can write the files, and adding a header then yeah.

dutow commented on contrib/pg_tde/src/keyring/keyring_vault.c:
```diff
@@ -300,6 +302,32 @@ get_key_by_name(GenericKeyring *keyring, const char *key_name, KeyringReturnCode
+	long		httpCode = 0;
+
+	/*
+	 * Just try to connect to the mount URL with an empty key name to see if
```
We could use the list operation (https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2#list-secrets, `/:secret-mount-path/metadata/:path`), and also verify the http status. The result isn't important, but it's a good practice to try to avoid incorrect requests, those can show up in the vault error log.

AndersAstrand replied:
Yeah it was just my lack of knowledge about vault showing. I hoped someone would have a better idea!

AndersAstrand replied:
Hmm.. I don't know if that's better since `:path` is required there as well, and we don't have a path to any secret or anything yet.

AndersAstrand replied:
I pushed a fixup commit to address this. Please have a look

AndersAstrand replied:
Unfortunately there doesn't seem to be a way to distinguish an invalid mount point from a valid one with no secrets yet. So currently we accept both 200 and 404 as valid statuses.

Unless we use `http://127.0.0.1:8200/v1/sys/mounts/secret`, do you know if a client is generally allowed to read the information under /sys/ if they have access to a mount point?

jeltz commented on contrib/pg_tde/src/keyring/keyring_file.c:
```diff
@@ -143,3 +145,19 @@ set_key_by_name(GenericKeyring *keyring, KeyInfo *key)
+	{
+		ereport(ERROR,
+				errcode_for_file_access(),
+				errmsg("Failed to open keyring file %s :%m", file_keyring->file_name));
```
The space is on the wrong side of the colon.

jeltz commented on contrib/pg_tde/src/keyring/keyring_kmip.c:
```diff
@@ -204,3 +206,15 @@ get_key_by_name(GenericKeyring *keyring, const char *key_name, KeyringReturnCode
 }
+
+static void
+validate(GenericKeyring *keyring)
```
There is a ping command in the protcol but I do not think our library supports it.

dutow replied:
also in this case at least establishing the connection should verify that the keys we have are okay.

jeltz replied:
Yeah, I think this is fine. If that library had supported the ping command it would have been worth using, but now: no.

Commit:
Fix error message in keyring_file.c

The colon was on the wrong side of the space.

Commit:
Rename KeyringProviderXLRecord

The new name, KeyringProviderRecordInFile, describes what it is rather
than what it's used for. But the real reason is that I want to use it
for other things than the WAL in future commits.

AndersAstrand commented on contrib/pg_tde/src/catalog/tde_keyring.c:
```diff
@@ -548,34 +449,191 @@ write_key_provider_info(KeyringProviderRecord *provider, Oid database_id,
+	close(fd);
+
+	/*
+	 * TODO: Should we use overflow aware integer math here? (and then also
```
@jeltz What do you think about this?

jeltz replied:
I say either throw an error if we run out of 32-bit IDs or use a 64-bit ID so we never can run out. Either is fine but this comment is just annoying. And 32-bit is fine becuase if we have ran out of that we have a huge file and thigns probably slow and horrible anyway.

AndersAstrand replied:
Yeah the comment was never meant to live past the review phase :). I throw an error now instead.

jeltz reviewed (approved):
Looks good!
